src/librustc_passes/ast_validation.rs

```diff
@@ -20,9 +20,79 @@ use errors::Applicability;
 
 struct AstValidator<'a> {
     session: &'a Session,
+
+    // Used to ban nested `impl Trait`, e.g., `impl Into<impl Debug>`.
+    // Nested `impl Trait` _is_ allowed in associated type position,
+    // e.g `impl Iterator<Item=impl Debug>`
```

Comment:
Maybe also add a note about why we do this (cause we are not sure whether we want it to be higher rank: `impl for<T: Debug> Into<T>`, or rank-1).